### PR TITLE
Add initial support for `tracing` and `tracing-perfetto`

### PR DESCRIPTION
**Pull request by atbrakhi**
This PR introduces the initial integration of `tracing` and `tracing-perfetto`. With this in place, we should be able to land and work on adding tracing events and logging in parallel

Co-authored-by: Delan Azabani <dazabani@igalia.com>
Signed-off-by: Rakhi Sharma <atbrakhi@igalia.com>

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors

### CODE REVIEW

File: ports/servoshell/desktop/cli.rs [hunk 1]
```diff
@@ -8,13 +8,32 @@ use getopts::Options;
...
+    // compositor-to-constellation and script-to-constellation logging.
+    // <https://tokio.rs/tokio/topics/tracing>
+    // <https://docs.rs/tracing-subscriber/0.3.18/tracing_subscriber/fmt/index.html#composing-layers>
+    let env_filter_layer = EnvFilter::from_default_env();
+    let fmt_layer = tracing_subscriber::fmt::layer();
+    tracing_subscriber::registry()
+        .with(env_filter_layer)
+        .with(fmt_layer)
+        .with(perfetto_layer)
+        .init();
```
Comment (mrobinson) [inline ports/servoshell/desktop/cli.rs:35]: Should this be turned on by a command-line argument which accepts an output filename?

Comment (atbrakhi) [reply ports/servoshell/desktop/cli.rs:35]: Oh, I have been thinking about this, but my concern comes from a different reason. This is really slowing down the build. On Mac, it takes around 5 minutes to build.

I've been discussing adding a flag with @delan and will wait for her feedback as well.

Comment (delan) [reply ports/servoshell/desktop/cli.rs:35]: yeah, there’s a smaller but still significant regression in build time on linux too (before [135s](https://github.com/user-attachments/files/16761228/cargo-timing-20240827T110216Z.html.gz), after [160s](https://github.com/user-attachments/files/16761233/cargo-timing-20240827T105843Z.html.gz)). what if we gate this behind a `tracing` or `perfetto` feature for now?

Comment (atbrakhi) [reply ports/servoshell/desktop/cli.rs]: sounds good to me. I have added the `tracing` feature flag in [519e954](https://github.com/servo/servo/pull/33188/commits/519e954d58413b8f9513d30a79ae694a48bc0c1f). I guess moving forward, we can modify it if needed.

File: ports/servoshell/desktop/cli.rs [hunk 2]
```diff
@@ -8,13 +8,39 @@ use getopts::Options;
...
+    {
+        let file = std::fs::File::create("servo.pftrace").unwrap();
+        let perfetto_layer = PerfettoLayer::new(std::sync::Mutex::new(file));
+        // Set up a custom tracing subscriber that behaves like the default fmt
+        // subscriber, but also lets us insert our own layers for things like
+        // compositor-to-constellation and script-to-constellation logging.
+        // <https://tokio.rs/tokio/topics/tracing>
+        // <https://docs.rs/tracing-subscriber/0.3.18/tracing_subscriber/fmt/index.html#composing-layers>
+        let env_filter_layer = EnvFilter::from_default_env();
+        let fmt_layer = tracing_subscriber::fmt::layer();
```
Comment (delan) [inline ports/servoshell/desktop/cli.rs:36]: i think these can be removed if we’re just using tracing for perfetto right now?

File: ports/servoshell/desktop/cli.rs [hunk 3]
```diff
@@ -8,13 +8,39 @@ use getopts::Options;
...
+        tracing_subscriber::registry()
+            .with(env_filter_layer)
+            .with(fmt_layer)
```
Comment (delan) [inline ports/servoshell/desktop/cli.rs:39]: same goes for these

File: ports/servoshell/Cargo.toml [hunk 4]
```diff
@@ -108,6 +109,9 @@ raw-window-handle = "0.6"
 shellwords = "1.0.0"
 surfman = { workspace = true, features = ["sm-x11", "sm-raw-window-handle-06"] }
 tinyfiledialogs = "3.0"
+tracing = { workspace = true, optional = true }
+tracing-subscriber = { workspace = true, features = ["env-filter"], optional = true }
```
Comment (delan) [inline ports/servoshell/Cargo.toml:113]: same goes for the extra feature

**Review by delan — CHANGES_REQUESTED** [review]
This panics on startup with:

```
$ ./mach build -d --features tracing
$ ./mach run -d
```

I think this is related to the [`log` compatibility](https://docs.rs/tracing/0.1.40/tracing/#consuming-log-records) that tracing [enables by default](https://docs.rs/tracing-subscriber/0.3.18/tracing_subscriber/index.html#optional-dependencies)? It seems to happen even if we remove the EnvFilter and fmt layers, as long as we init the registry, so we may need to disable the feature.

**Comment by atbrakhi** [comment]
> This panics on startup with:
> 
> ```
> $ ./mach build -d --features tracing
> $ ./mach run -d
> ```
> 
> I think this is related to the [`log` compatibility](https://docs.rs/tracing/0.1.40/tracing/#consuming-log-records) that tracing [enables by default](https://docs.rs/tracing-subscriber/0.3.18/tracing_subscriber/index.html#optional-dependencies)? It seems to happen even if we remove the EnvFilter and fmt layers, as long as we init the registry, so we may need to disable the feature.

@delan as you said even removing EnvFilter and fmt layers are causing panic. How do you feel about using `set_global_default()`? It should help us avoid panic. I have pushed this change in [a298648](https://github.com/servo/servo/pull/33188/commits/a2986488412812ac9775147752e3efbedf8645f0), and have been testing this changes with #33188

**Comment by delan** [comment]
> How do you feel about using `set_global_default()`? It should help us avoid panic.

Interesting! That does seem to be exactly the difference between [set_global_default](https://docs.rs/tracing/0.1.40/src/tracing/subscriber.rs.html#39-44) and [SubscriberInitExt::init](https://docs.rs/tracing-subscriber/0.3.18/src/tracing_subscriber/util.rs.html#58-92). I think that’s ok if we add a comment explaining it, though it may be simpler to disable the `tracing-log` feature. I also think we should still remove the EnvFilter and fmt layers, unless we need them for profiling.

**Comment by delan** [comment]
(rebased to fix my email address; no actual changes)

**Comment by atbrakhi** [comment]
> > How do you feel about using `set_global_default()`? It should help us avoid panic.
> 
> Interesting! That does seem to be exactly the difference between [set_global_default](https://docs.rs/tracing/0.1.40/src/tracing/subscriber.rs.html#39-44) and [SubscriberInitExt::init](https://docs.rs/tracing-subscriber/0.3.18/src/tracing_subscriber/util.rs.html#58-92). I think that’s ok if we add a comment explaining it, though it may be simpler to disable the `tracing-log` feature. I also think we should still remove the EnvFilter and fmt layers, unless we need them for profiling.

Sorry, should have mentioned in previous comment, even when I remove the `tracing-log` feature and also the EnvFilter and fmt layers, it still crashes. that's how i ended up with `set_global_default`

**Comment by delan** [comment]
> Sorry, should have mentioned in previous comment, even when I remove the `tracing-log` feature and also the EnvFilter and fmt layers, it still crashes, that's how i ended up with `set_global_default`

No worries, in that case `set_global_default` sounds good.

**Comment by delan** [comment]
I noticed that `tracing-log` still appears in `cargo tree -e features` if we specify `default-features = false` in the servoshell Cargo.toml, but disappears if we specify it in the workspace `[dependencies]`:

```toml
# Cargo.toml
tracing-subscriber = { version = "0.3.18", default-features = false }

# ports/servoshell/Cargo.toml
tracing-subscriber = { workspace = true }
```

This seems to be documented [in the Cargo book](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table):

> Specifying a workspace dependency is similar to [package dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html) except: […] [features](https://doc.rust-lang.org/cargo/reference/features.html) declared in this table are additive with the features from [dependencies]

File: ports/servoshell/Cargo.toml (hunk 4 above)
Comment (atbrakhi) [reply ports/servoshell/Cargo.toml:115]: we need `env-filter` later

**Review by delan — APPROVED** [review]

**Pull request merged**